jpyinterpreter-docs/src/modules/ROOT/pages/jpyinterpreter-architecture/jpyinterpreter-architecture.adoc

```diff
@@ -0,0 +1,125 @@
+It also provides the `translate_python_class_to_java_class` function, which converts a user supplied CPython class, convert it to a `PythonCompiledClass`, and pass it to `PythonClassTranslator` to translate the class.
+
+The Java components are:
+
+- `PythonBytecodeToJavaBytecodeTranslator`, which acts as the entrypoint for function translation.
+It is responsible for:
+
+** Setting up the `JavaPythonClassWriter` and `MethodVisitor` used for bytecode generation.
+
+** Creating and setting fields on the generated class objects (See <<_pythonbytecodetojavabytecodetranslator>> for details).
+
+** Do the leg work of moving/translating Java parameters (ex: `int`) into `PythonLikeObject`.
+
+** Setup cells variables.
+
+** Delegating to `PythonGeneratorTranslator` when it detects the function being translated is a generator.
+
+** Using `FlowGraph` to calculate the `StackMetadata` of each instruction.
+
+** Calling the `implement` method on every opcode in the `Opcode` list with the `StackMetadata` for the opcode and the `FunctionMetadata` for the overall function.
+
+- `PythonGeneratorTranslator` is like `PythonBytecodeToJavaBytecodeTranslator` but for generators.
+It breaks a single generator function into multiple `advance` functions, and generate each `advance` function bytecode independently.
+
+- `FlowGraph`, which calculates the `StackMetadata` that corresponds to each `Opcode`.
+It is responsible for unifying the `StackMetadata` from all jump sources for each `Opcode` that is a jump target.
+For instance, if two sources with the same target have post `StackMetadata` of `... int` and `... bool` respectively, `FlowGraph` will unify that to `... int` (since `bool` is a subclass of `int`, for better or worse).
+
+- `StackMetadata` stores metadata about the stack and local variables.
+Each `Opcode` get its own `StackMetadata` instance.
+It is mostly used to perform optimizations; for instance, if we detect the top two items on the stack are `int` and `int` for the `BINARY_ADD` instruction, we can change the (normally complex due to Python semantics) `BINARY_ADD` bytecode into a single method call.
+
+- `FunctionMetadata` stores metadata about the function (for instance, the `MethodVisitor` to use to generate bytecode). Each `Opcode` get the same `FunctionMetadata` instance.
+
+- `Opcode` are the interface between CPython opcodes and the `Implementors`.
+Each describe a particular operation, and are usually (but not always) correspond to a CPython opcode.
+Some CPython opcodes will map to the same `Opcode` implementation.
+
+- `Implementors` are responsible for generating the Java bytecode corresponding to CPython bytecode.
+They can be found in the `implementors` package.
+
+In a diagram, the overall process of compiling a function looks like this:
+
+image::jpyinterpreter-architecture/jpyinterpreter-architecture.png[A diagram showing how JPyInterpreter classes interact]
+
+
+== Types
```

Comment:
I'm wondering if some of this chapter shouldn't rather be pointers to Javadoc.
That way, we reduce the likelihood of this information (such as field and method names) becoming stale.